postgis/lwgeom_accum.c

```diff
@@ -114,64 +87,105 @@ pgis_geometry_accum_transfn(PG_FUNCTION_ARGS)
 
+		for (int i = 0; i < n; i++)
+		{
+			Datum argument = PG_GETARG_DATUM(i+2);
+			Oid dataOid = get_fn_expr_argtype(fcinfo->flinfo, i+2);
+			state->data[i] = datumCopy(argument, get_typbyval(dataOid), get_typlen(dataOid));
 		}
-		MemoryContextSwitchTo(old);
 	}
 	else
 	{
-		p = (pgis_abs*) PG_GETARG_POINTER(0);
+		state = (CollectionBuildState*) PG_GETARG_POINTER(0);
 	}
-	state = p->a;
-	elem = PG_ARGISNULL(1) ? (Datum) 0 : PG_GETARG_DATUM(1);
-	state = accumArrayResult(state,
-	                         elem,
-	                         PG_ARGISNULL(1),
-	                         arg1_typeid,
-	                         aggcontext);
-	p->a = state;
-
-	PG_RETURN_POINTER(p);
-}
 
+	if (!PG_ARGISNULL(1))
+		gser = PG_GETARG_GSERIALIZED_P(1);
 
+	/* Take a copy of the geometry into the aggregate context */
+	old = MemoryContextSwitchTo(aggcontext);
+	if (gser)
+		geom = lwgeom_clone_deep(lwgeom_from_gserialized(gser));
+
+	/* Initialize or append to list as necessary */
+	if (state->geoms)
+		state->geoms = lappend(state->geoms, geom);
+	else
+		state->geoms = list_make1(geom);
 
-Datum pgis_accum_finalfn(pgis_abs *p, MemoryContext mctx, FunctionCallInfo fcinfo);
+	MemoryContextSwitchTo(old);
+
+	PG_RETURN_POINTER(state);
+}
+
+
+Datum pgis_accum_finalfn(CollectionBuildState *state, MemoryContext mctx, FunctionCallInfo fcinfo);
 
 /**
 ** The final function rescues the built array from the side memory context
 ** using the PostgreSQL built-in function makeMdArrayResult
 */
 Datum
-pgis_accum_finalfn(pgis_abs *p, MemoryContext mctx, __attribute__((__unused__)) FunctionCallInfo fcinfo)
+pgis_accum_finalfn(CollectionBuildState *state, MemoryContext mctx, __attribute__((__unused__)) FunctionCallInfo fcinfo)
 {
+	ListCell *l;
+	size_t nelems = 0;
+	Datum *elems;
+	bool *nulls;
+	int16 elmlen;
+	bool elmbyval;
+	char elmalign;
+	int i = 0;
+	ArrayType *arr;
 	int dims[1];
-	int lbs[1];
-	ArrayBuildState *state;
-	Datum result;
+	int lbs[1] = {1};
 
 	/* cannot be called directly because of internal-type argument */
 	Assert(fcinfo->context &&
 	       (IsA(fcinfo->context, AggState) ||
 	        IsA(fcinfo->context, WindowAggState))
 	       );
 
-	state = p->a;
-	dims[0] = state->nelems;
-	lbs[0] = 1;
-	result = makeMdArrayResult(state, 1, dims, lbs, mctx, false);
-	return result;
+	/* Retrieve geometry type metadata */
+	get_typlenbyvalalign(state->geomOid, &elmlen, &elmbyval, &elmalign);
+	nelems = list_length(state->geoms);
+
+	/* Build up an array, because that's what we pass to all the */
+	/* specific final functions */
+	elems = palloc(nelems * sizeof(Datum));
+	nulls = palloc(nelems * sizeof(bool));
+
+	foreach (l, state->geoms)
+	{
+		LWGEOM *geom = (LWGEOM*)(lfirst(l));
+		Datum elem = (Datum)0;
+		bool isNull = true;
+		if (geom)
+		{
+			GSERIALIZED *gser = geometry_serialize(geom);
+			elem = PointerGetDatum(gser);
+			isNull = false;
+		}
+		elems[i] = elem;
+		nulls[i] = isNull;
+		i++;
+
+		if (i >= nelems)
```

Comment:
Is the extra defense barrier here necessary?

Reply:
Belt and suspenders? I just don't want to get punked.